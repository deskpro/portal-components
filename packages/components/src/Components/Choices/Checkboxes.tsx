import * as React from 'react';
import classNames from 'classnames';
import { FieldArray, getIn } from 'formik';
import Field, { FieldProps, FieldState } from '../Field';

interface CheckboxesProps extends FieldProps {
  options: {
    label: string | React.ReactElement;
    value: any;
  }[];
}

interface CheckboxesState extends FieldState {
  focused: boolean;
}

class Checkboxes extends Field<CheckboxesProps, CheckboxesState> {
  constructor(props) {
    super(props);
    this.state = {
      focused: false,
    };
  }

  handleKeyPress = (e, form, arrayHelpers, value) => {
    const { name } = this.props;
    if (e.key === ' ') {
      e.preventDefault();
      if (!form.values[name].includes(value)) {
        arrayHelpers.push(value);
      } else {
        const idx = getIn(form.values, name).indexOf(value);
        arrayHelpers.remove(idx);
      }
    }
  };

  handleBlur = () => {
    this.setState({
      focused: false
    });
  };

  handleFocus = () => {
    this.setState({
      focused: true
    });
  };

  onChange = (e, form, name, value, arrayHelpers) => {
    e.preventDefault();
    const idx = getIn(form.values, name).indexOf(value);
    if (idx === -1) {
      arrayHelpers.push(value);
    } else {
      arrayHelpers.remove(idx);
    }
  };

  renderLabel = () => {
    const {
      label,
    } = this.props;
    const htmlFor = this.id;
    return (
      <label className="dp-pc_checkboxes_label" htmlFor={htmlFor}>
        {label}
      </label>
    );
  };

  renderDescription = () => {
    const { description } = this.props;

    if (description) {
      return <span className="dp-pc_description dp-pc_checkboxes_description">{description}</span>;
    }
    return null;
  }

  renderField = ({ form }) => {
    const {
      name, label, options, className, fClassName, errorsName, i18n, ...props
    } = this.props;
    return (
      <FieldArray
        name={name}
        render={arrayHelpers => (
          <div
            className={classNames('dp-pc_checkboxes', { focused: this.state.focused }, className)}
            {...props}
          >
            {options.map(option => (
              <label
                key={option.value}
                htmlFor={`${this.id}-${option.value}`}
                tabIndex={0}
                className="dp-pc_checkboxes_checkbox"
                onKeyPress={e => this.handleKeyPress(e, form, arrayHelpers, option.value)}
                onFocus={this.handleFocus}
                onBlur={this.handleBlur}
                onClick={(e) => { this.onChange(e, form, name, option.value, arrayHelpers); }}
              >
                <input
                  id={`${this.id}-${option.value}`}
                  name={name}
                  type="checkbox"
                  value={option.value}
                  checked={!!form.values[name] && form.values[name].includes(option.value)}
                  onChange={(e) => { this.onChange(e, form, name, option.value, arrayHelpers); }}
                />
                <i onClick={(e) => { this.onChange(e, form, name, option.value, arrayHelpers); }} />
                {' '}
                <span className="label">
                  {option.label}
                </span>
              </label>
            ))}
          </div>
        )}
      />
    );
  }
}

export default Checkboxes;
