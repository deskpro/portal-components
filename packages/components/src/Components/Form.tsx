import * as React from 'react';
import { Form as FormikForm } from 'formik';
import classNames from 'classnames';

interface IProps {
  className?: string;
  showHover?: boolean;
  children:   React.ReactNode;
}

const Form = ({
  className = '',
  showHover = true,
  ...props
}: IProps) => (
  <FormikForm
    className={classNames('dp-pc_form', className, { 'show-hover': showHover })}
    {...props}
  />);



export default Form;
