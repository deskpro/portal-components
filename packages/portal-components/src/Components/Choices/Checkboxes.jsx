import React from 'react';
import PropTypes from 'prop-types';
import { FieldArray, getIn } from 'formik';
import Field from '../Field';

class Checkboxes extends Field {
  constructor(props) {
    super(props);
    this.type = 'checkbox';
  }

  renderField = (form) => {
    const {
      name, id, label, options, ...props
    } = this.props;
    return (
      <FieldArray
        name={name}
        render={arrayHelpers => (
          <div {...props}>
            {options.map(option => (
              <div key={option.value}>
                <label htmlFor={`${id}-${option.value}`}>
                  <input
                    id={`${id}-${option.value}`}
                    name={name}
                    type="checkbox"
                    value={option.value}
                    checked={form.values[name].includes(option.value)}
                    onChange={(e) => {
                      if (e.target.checked) arrayHelpers.push(option.value);
                      else {
                        const idx = getIn(form.values, name).indexOf(option.value);
                        arrayHelpers.remove(idx);
                      }
                    }}
                  />{' '}
                  {option.label}
                </label>
              </div>
            ))}
          </div>
        )}
      />
    );
  }
}

Checkboxes.propTypes = Object.assign(Field.propTypes, {
  options: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
    value: PropTypes.any.isRequired
  }))
});

export default Checkboxes;
