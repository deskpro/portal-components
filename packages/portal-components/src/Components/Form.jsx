import React from 'react';
import { Form as FormikForm } from 'formik';

const Form = ({
  ...props
}) => <FormikForm {...props} />;

export default Form;
