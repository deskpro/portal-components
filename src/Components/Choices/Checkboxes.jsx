import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { FieldArray, getIn } from 'formik';
import Field from '../Field';

class Checkboxes extends Field {
  constructor(props) {
    super(props);
    this.state = {
      focused: false,
    };
  }

  handleKeyPress = (e, form, arrayHelpers, value) => {
    const { name } = this.props;
    if (e.key === ' ') {
      e.preventDefault();
      if (!form.values[name].includes(value)) {
        arrayHelpers.push(value);
      } else {
        const idx = getIn(form.values, name).indexOf(value);
        arrayHelpers.remove(idx);
      }
    }
  };

  handleBlur = () => {
    this.setState({
      focused: false
    });
  };

  handleFocus = () => {
    this.setState({
      focused: true
    });
  };

  renderLabel = () => {
    const {
      label,
    } = this.props;
    const htmlFor = this.id;
    /* eslint-disable jsx-a11y/label-has-for */
    return (
      <label className="dp-pc_checkboxes_label" htmlFor={htmlFor}>
        {label}
      </label>
    );
    /* eslint-enable jsx-a11y/label-has-for */
  };
  renderField = (form) => {
    const {
      name, label, options, className, ...props
    } = this.props;
    /* eslint-disable jsx-a11y/no-noninteractive-tabindex */
    /* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
    return (
      <FieldArray
        name={name}
        render={arrayHelpers => (
          <div
            className={classNames('dp-pc_checkboxes', { focused: this.state.focused }, className)}
            {...props}
          >
            {options.map(option => (
              <label
                key={option.value}
                htmlFor={`${this.id}-${option.value}`}
                tabIndex="0"
                className="dp-pc_checkboxes_checkbox"
                onKeyPress={e => this.handleKeyPress(e, form, arrayHelpers, option.value)}
                onFocus={this.handleFocus}
                onBlur={this.handleBlur}
              >
                <input
                  id={`${this.id}-${option.value}`}
                  name={name}
                  type="checkbox"
                  value={option.value}
                  checked={form.values[name].includes(option.value)}
                  onChange={(e) => {
                    if (e.target.checked) arrayHelpers.push(option.value);
                    else {
                      const idx = getIn(form.values, name).indexOf(option.value);
                      arrayHelpers.remove(idx);
                    }
                  }}
                />
                <i />
                {' '}
                <span className="label">
                  {option.label}
                </span>
              </label>
            ))}
          </div>
        )}
      />
    );
    /* eslint-enable jsx-a11y/no-noninteractive-tabindex */
    /* eslint-enable jsx-a11y/no-noninteractive-element-interactions */
  }
}

Checkboxes.propTypes = {
  ...Field.propTypes,
  options: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
    value: PropTypes.any.isRequired
  })),
};

export default Checkboxes;
