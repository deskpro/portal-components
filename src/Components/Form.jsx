import React from 'react';
import { Form as FormikForm } from 'formik';

const Form = ({
  ...props
}) => {
  return <FormikForm {...props} />
};

export default Form;