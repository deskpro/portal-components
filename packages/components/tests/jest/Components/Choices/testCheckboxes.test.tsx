import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { withFormik } from 'formik';
import { describe, expect, it } from '@jest/globals';
import Form from '../../../../src/Components/Form';
import Checkboxes from '../../../../src/Components/Choices/Checkboxes';

interface FormValues {
  checkboxes: string[];
}

interface MyFormikProps {
  checkboxes: string[];
}

describe('>>> Checkboxes --- Snapshot', () => {
  it('+++capturing Snapshot of Checkboxes', () => {
    const options = [
      { value: 'first', label: 'First' },
      { value: 'second', label: 'Second' },
      { value: 'last', label: 'Last' },
    ];
    const App = () => (
      <Form>
        <Checkboxes name="checkboxes" options={options} />
      </Form>
    );
    const FormikApp = withFormik<MyFormikProps, FormValues>({
      mapPropsToValues({
        checkboxes
      }) {
        return {
          checkboxes: checkboxes || []
        };
      },
      handleSubmit(values, { setSubmitting }) {
        setSubmitting(false);
      }
    })(App);
    const renderedValue = renderer.create(<FormikApp checkboxes={[]} />).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });

  it('+++capturing Snapshot of Checkboxes with existing value', () => {
    const options = [
      { value: 'first', label: 'First' },
      { value: 'second', label: 'Second' },
      { value: 'last', label: 'Last' },
    ];
    const App = () => (
      <Form>
        <Checkboxes name="checkboxes" id="checkbox" options={options} />
      </Form>
    );
    const FormikApp = withFormik<MyFormikProps, FormValues>({
      mapPropsToValues({
        checkboxes
      }) {
        return {
          checkboxes: checkboxes || ['first']
        };
      },
      handleSubmit(values, { setSubmitting }) {
        setSubmitting(false);
      }
    })(App);
    const renderedValue = renderer.create(<FormikApp checkboxes={[]} />).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
});
