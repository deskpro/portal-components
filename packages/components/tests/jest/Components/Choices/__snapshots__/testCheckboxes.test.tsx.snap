// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> Checkboxes --- Snapshot +++capturing Snapshot of Checkboxes 1`] = `
<form
  action="#"
  className="dp-pc_form show-hover"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="dp-pc_field"
  >
    <label
      className="dp-pc_checkboxes_label"
      htmlFor="field_1"
    >
      
    </label>
    <div
      className="dp-pc_checkboxes"
      description=""
      id={null}
      required={false}
    >
      <label
        className="dp-pc_checkboxes_checkbox"
        htmlFor="field_1-first"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        tabIndex={0}
      >
        <input
          checked={false}
          id="field_1-first"
          name="checkboxes"
          onChange={[Function]}
          type="checkbox"
          value="first"
        />
        <i
          onClick={[Function]}
        />
         
        <span
          className="label"
        >
          First
        </span>
      </label>
      <label
        className="dp-pc_checkboxes_checkbox"
        htmlFor="field_1-second"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        tabIndex={0}
      >
        <input
          checked={false}
          id="field_1-second"
          name="checkboxes"
          onChange={[Function]}
          type="checkbox"
          value="second"
        />
        <i
          onClick={[Function]}
        />
         
        <span
          className="label"
        >
          Second
        </span>
      </label>
      <label
        className="dp-pc_checkboxes_checkbox"
        htmlFor="field_1-last"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        tabIndex={0}
      >
        <input
          checked={false}
          id="field_1-last"
          name="checkboxes"
          onChange={[Function]}
          type="checkbox"
          value="last"
        />
        <i
          onClick={[Function]}
        />
         
        <span
          className="label"
        >
          Last
        </span>
      </label>
    </div>
  </div>
</form>
`;

exports[`>>> Checkboxes --- Snapshot +++capturing Snapshot of Checkboxes with existing value 1`] = `
<form
  action="#"
  className="dp-pc_form show-hover"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="dp-pc_field"
  >
    <label
      className="dp-pc_checkboxes_label"
      htmlFor="checkbox"
    >
      
    </label>
    <div
      className="dp-pc_checkboxes"
      description=""
      id="checkbox"
      required={false}
    >
      <label
        className="dp-pc_checkboxes_checkbox"
        htmlFor="checkbox-first"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        tabIndex={0}
      >
        <input
          checked={false}
          id="checkbox-first"
          name="checkboxes"
          onChange={[Function]}
          type="checkbox"
          value="first"
        />
        <i
          onClick={[Function]}
        />
         
        <span
          className="label"
        >
          First
        </span>
      </label>
      <label
        className="dp-pc_checkboxes_checkbox"
        htmlFor="checkbox-second"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        tabIndex={0}
      >
        <input
          checked={false}
          id="checkbox-second"
          name="checkboxes"
          onChange={[Function]}
          type="checkbox"
          value="second"
        />
        <i
          onClick={[Function]}
        />
         
        <span
          className="label"
        >
          Second
        </span>
      </label>
      <label
        className="dp-pc_checkboxes_checkbox"
        htmlFor="checkbox-last"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyPress={[Function]}
        tabIndex={0}
      >
        <input
          checked={false}
          id="checkbox-last"
          name="checkboxes"
          onChange={[Function]}
          type="checkbox"
          value="last"
        />
        <i
          onClick={[Function]}
        />
         
        <span
          className="label"
        >
          Last
        </span>
      </label>
    </div>
  </div>
</form>
`;
